### PR TITLE
Field refine

### PR DESCRIPTION
**Pull request by rochSmets**
This PR is associated to issue #412

### CODE REVIEW

File: tests/amr/data/field/refine/test_field_refine.cpp [hunk 1]
```diff
@@ -39,523 +38,360 @@ TEST(UniformIntervalPartition, givesCorrectPartitionsForPrimalEven)
...
-TEST(AFieldLinearRefineIndexesAndWeights1D, giveACorrectStartIndexForPrimalOddRatio)
+TYPED_TEST(aFieldLinearRefineIndexesAndWeights, giveACorrectStartIndexForPrimalQty)
 {
-    std::size_t constexpr dimension{1};
-    std::array<QtyCentering, dimension> centering{{QtyCentering::primal}};
-    SAMRAI::hier::IntVector ratio{SAMRAI::tbox::Dimension{dimension}, 9};
-    FieldRefineIndexesAndWeights<dimension> indexesAndWeights{centering, ratio};
+    static constexpr auto dim = TypeParam{}();
+
+    std::array<QtyCentering, dim> centering = makeCentering<dim>(QtyCentering::primal);
```
[the remark this answers is not in the document]
Comment (rochSmets) [reply tests/amr/data/field/refine/test_field_refine.cpp]: yes, sure : done,

File: tests/amr/data/field/refine/test_field_refine.cpp [hunk 2]
```diff
@@ -39,523 +38,360 @@ TEST(UniformIntervalPartition, givesCorrectPartitionsForPrimalEven)
...
+    SAMRAI::hier::IntVector ratio{SAMRAI::tbox::Dimension{dim}, 2};
+    FieldRefineIndexesAndWeights<dim> indexesAndWeights{centering, ratio};
+
+    std::array<Point<int, dim>, 4> fineIndexes          = makeArrayOfPoints<dim, 4>({0, 1, 2, 6});
```
[the remark this answers is not in the document]
Comment (rochSmets) [reply tests/amr/data/field/refine/test_field_refine.cpp:175]: these arrays are now `constexpr`

[the remark this answers is not in the document]
Comment (rochSmets) [reply tests/amr/data/field/refine/test_field_refine.cpp:175]: negative fine indexes are now included,

File: tests/amr/data/field/refine/test_field_refine.cpp [hunk 3]
```diff
@@ -39,523 +38,360 @@ TEST(UniformIntervalPartition, givesCorrectPartitionsForPrimalEven)
...
+    SAMRAI::hier::IntVector ratio{SAMRAI::tbox::Dimension{dim}, 2};
+    FieldRefineIndexesAndWeights<dim> indexesAndWeights{centering, ratio};
 
-TEST(AFieldLinearRefineIndexesAndWeights1D, giveACorrectStartIndexForDualOddRatio)
-{
-    std::size_t constexpr dimension{1};
-    std::array<QtyCentering, dimension> centering{{QtyCentering::dual}};
-    SAMRAI::hier::IntVector ratio{SAMRAI::tbox::Dimension{dimension}, 7};
-    FieldRefineIndexesAndWeights<dimension> indexesAndWeights{centering, ratio};
+    std::array<Point<int, dim>, 4> fineIndexes          = makeArrayOfPoints<dim, 4>({0, 1, 2, 6});
```
[the remark this answers is not in the document]
Comment (rochSmets) [reply tests/amr/data/field/refine/test_field_refine.cpp:218]: done,

File: tests/amr/data/field/refine/test_field_refine.cpp [hunk 4]
```diff
@@ -39,523 +38,360 @@ TEST(UniformIntervalPartition, givesCorrectPartitionsForPrimalEven)
...
-    EXPECT_DOUBLE_EQ(xWeights[4][1], 4 * smallCellSize);
-    EXPECT_DOUBLE_EQ(xWeights[4][0], 1. - 4 * smallCellSize);
-
-    EXPECT_DOUBLE_EQ(xWeights[5][1], 5 * smallCellSize);
-    EXPECT_DOUBLE_EQ(xWeights[5][0], 1. - 5 * smallCellSize);
-}
-
-
-
+    EXPECT_DOUBLE_EQ(xWeights[1][1], 0.5);
```
[the remark this answers is not in the document]
Comment (rochSmets) [reply tests/amr/data/field/refine/test_field_refine.cpp:259]: yes of course... easier to read ; done,

File: tests/amr/data/field/refine/test_field_refine.cpp [hunk 5]
```diff
@@ -39,523 +38,360 @@ TEST(UniformIntervalPartition, givesCorrectPartitionsForPrimalEven)
...
+            EXPECT_EQ(expectedWeightIndex, zWeight);
+        }
+    }
 }
 
 
-
-
-TEST(AFieldLinearIndexesAndWeights1D, giveACorrectiWeightForDualOddRatio)
+TYPED_TEST(aFieldLinearRefineIndexesAndWeights, giveACorrectWeightForDualQty)
```
[the remark this answers is not in the document]
Comment (rochSmets) [reply tests/amr/data/field/refine/test_field_refine.cpp]: done,

**Review by nicolasaunai — APPROVED** [review]

**Pull request merged**